### ISSUE TITLE
Lint using a variable name the same as a type

### ISSUE THREAD

**Issue by shepmaster** [rust-lang-nursery/rust-clippy#2053]
Some examples from Rust 1.20:

🙅 — compiler generates an error (can't shadow unit structs)

```rust
struct Dog;
fn x(Dog: Dog) {}
```

⚠️ — compiler generates a warning (non-snake case variable)

```rust
struct Dog {}
fn x(Dog: Dog) {}
```

😖  — compiler doesn't generate a warning (and indeed the variable `str` is valid to use)

```rust
struct Dog {}
fn x(str: Dog) {}
```

----

I propose that Clippy warn when a variable name matches an in-scope type. I'm not even sure if it's possible to know all the types, but using the built-in lowercase primitive types (e.g. `bool`, `str`, `i32`) would be a great start.

**Comment by shepmaster** [comment]
This should apply to lifetime names as well.
